src/Domains/Domains.jl

```diff
@@ -1,56 +1,64 @@
 module Domains
 
-import ..Geometry
+import ..Geometry: Geometry, float_type
 using IntervalSets
 export RectangleDomain
 
+"""
+    AbstractDomain
 
+A domain represents a region of space.
+"""
 abstract type AbstractDomain end
 
-abstract type HorizontalDomain <: AbstractDomain end
-abstract type VerticalDomain <: AbstractDomain end
-
 const BCTagType = Union{Nothing, Tuple{Symbol, Symbol}}
 
-struct IntervalDomain{CT, B} <: VerticalDomain where {
+float_type(domain::AbstractDomain) = float_type(coordinate_type(domain))
+
+"""
+    boundary_names(obj::Union{AbstractDomain, AbstractMesh, AbstractTopology})
+
+The boundary names of a spatial domain. This is a tuple of `Symbol`s.
+"""
+function boundary_names end
+
+struct IntervalDomain{CT, B} <: AbstractDomain where {
     CT <: Geometry.Abstract1DPoint{FT},
     B <: BCTagType,
 } where {FT}
     coord_min::CT
     coord_max::CT
-    boundary_tags::B
+    boundary_names::B
 end
 
+isperiodic(domain::IntervalDomain) = isnothing(domain.boundary_names)
+boundary_names(domain::IntervalDomain) =
+    isperiodic(domain) ? () : domain.boundary_names
```

Comment:
should this return `nothing` for periodic domains?

Reply:
no, an empty set makes more sense (it's the collection of all boundary names)

Reply:
maybe it should be `unique(domain.boundary_names)` in the other case though.

Reply:
if it is a set interpretation that would be a good change, or you could just check upon construction that the names are unique which seems better / less confusing

Reply:
It is now unique.